---
title: "[2. Node.js 필수 패턴] 2.1 Callback Pattern"
date: 2019-05-07 09:46:00
categories: nodejs designpattern
comments: true

---

## 2.1 Callback Pattern

콜백은 리액터 패턴 핸들러를 구현한 것이다. 또한 Node.js에 독특한 프로그래밍 스타일을 제공하는 상징 중 하나이다. 콜백은 작업 결과를 전달하기 위해 호출되는 함수이며, 비동기 작업을 처리할 때 반드시 필요하다. 이들은 항상 동기적으로 실행되는 return 명령의 사용을 대신한다. JavaScript는 콜백을 표현할 수 있는 훌륭한 언어이다. 함수가 일급 클래스 객체(first class object)여서 변수에 쉽게 할당하거나, 인수로 전달되거나, 다른 함수 호출에서 반환되거나 자료구조에 저장될 수 있기 때문이다. 콜백을 구현하는 또 다른 이상적인 구조는 클로저(Closures)이다. 클로저를 사용하면 실제로 함수가 작성된 환경을 참조할 수 있다. 콜백이 언제 어디서 호출되는 지에 관계없이 비동기 작업이 요청된 커텍스트를 항상 유지할 수 있기 때문이다.

`` 클로저에 대한 설명 참고 :  https://opentutorials.org/course/743/6544 ``

### 2.1.1 연속 전달 방식(The Continuation-Passing Style)

- JavaScript에서 콜백은 다른 함수에 인수로 전달되는 함수이며, 작업이 완료되면 결과로 호출된다. 함수형 프로그래밍에서 결과를 전달하는 방식을 **연속 전달 방식(CPS)**이라고 한다.
- **연속 전달 방식(CPS)** 은 일반적인 개념으로, 항상 비동기 작업과 관련이 있는 것은 아니다. 단순히 결과를 호출자에게 직접 반환하는 대신 다른 함수(콜백)로 전달함으로써 결과를 전달하는 것을 말한다. 

### **동기식 연속 전달 방식**

```javascript
function add(a, b) {
	return a + b;
}
```

위는 결과를 return 문을 통해 호출자에게 전달하는 간단한 동기 함수를 나타낸다. 이것을 **직접 스타일(direct style)**이라고 하며, 동기화 프로그래밍에서 일반적으로 결과를 반환하는 방식을 보여준다.

```javascript
function add(a, b, callback) {
	callback(a + b);
}
```

위는 직접 스타일로 작성된 동기 함수를 연속 전달 방식으로 바꾼 코드이다. add() 함수는 동기화된 CPS 함수로 콜백이 완료될 때만 값을 반환한다.

아래 코드는 사용 방법을 보여준다.

```javascript
function add(a, b, callback) {
	callback(a + b);
}

console.log('before');
add(1, 2, result => { console.log('Result: ', + result) });
console.log('after');

======================= 출력 결과 =======================
before
result : 3
after
```

add()가 동기 함수이므로 순서대로 출력한다.

###**비동기식 연속 전달 방식**

```javascript
function additionAsync(a, b, callback) {
	setTimeout(() => callback(a + b), 100);
}

console.log('before');
additionAsync(1, 2, result => { console.log('Result: ', + result) });
console.log('after');


======================= 출력 결과 =======================
before
after
Result:  3
```

동기식으로 작성되었던 add() 함수를 비동기적으로 작동하는 예시를 보이기 위해 setTimeout()을 사용하여 콜백의 비동기 호출을 가정했다. additionAsync를 사용하면 출력 결과는 위와 같이 나온다. setTimeout()은 비동기 작업을 실행시키기 때문에 콜백의 실해이 끝날 때까지 기다리지 않는 대신, 즉시 반환되어 additionAsync()로 제어를 돌려주어 제어가 호출자에게 반환된다. Node.js의 이 속성은 비동기 요청이 전달된 후 즉시 제어를 이벤트 루프에 돌려주어 큐(대기열)에 있는 새로운 이벤트가 처리도리 수 있도록 하기 때문에 매우 중요하다.

작동방식은 아래와 같다.

![2_1_image1](https://firstwook.github.io/assets/img/2_1_image1.png)

비동기 작업이 완료되면 실행은 비동기 함수에 제공된 콜백에서부터 다시 계속된다. 실행은 **이벤트 루프**에서 시작되기 떄문에 새로운 스택을 갖는다. 클로저 덕분에 콜백이 다른 시점과 다른 위치에서 호출되더라도 비동기 함수의 호출자 컨텍스트를 유지하므로 JavaScript가 유용한 지점이다.

동기 함수는 조작을 완료할 때까지 블록한다. 비동기 함수는 제어를 즉시 반환하고 결과는 이벤트 루프의 다음 사이클에서 핸들러(이 경우에는 콜백)로 전달된다.

###**비 연속 전달(Non-continuation-passing) 방식의 콜백**

함수에 콜백 인자가 있으면, 함수가 비동기식이거나 연속 전달 스타일(CPS)을 사용한다고 가정할 수 있다. 그러나 항상 그런 것은 아니다. 

```javascript
const result = [1, 5, 7].map(element => element - 1);
console.log(result); //[ 0, 4, 6 ]
```

위의 예시와 같이 Array 객체의 map() 함수를 사용한 예를 보면, 콜백은 배열 내의 요소를 반복하는데 사용될 뿐 연산 결과를 전달하지 않는다. 실제 여기서 결과는 직접적인 방식으로 동기적으로 반환한다. 보통 사용하려는 함수에 대해, 해당 함수의 콜백의 목적은 API 문서에 분명하게 명시된다.

###2.1.2 동기냐? 비동기냐?

- 동기식이과 비동기식 둘 다 모두 피해야 할 것은 API의 이러한 특성과 관련하여 모순과 혼돈을 만드는 것이다.
- 그렇게 하면 발견하기 어렵고 재현이 불가능한 문제를 야기할 수 있다.

### **예측할 수 없는 함수**

가장 위험한 상황 중 하나는 특정 조건에서 동기적으로 동작하고 다른 조건에서는 비동기적으로 동작하는 API를 갖는 것이다.

```javascript
const fs = require('fs');
const cache = {};
function inconsistentRead(filename, callback){
  if(cache[filename]){
    //동기적으로 호출됨
    callback(cache[filename]);
  }else {
    //비동기함수
    fs.readFile(filename, 'utf8', (err, data) => {
      cache[filename] = data;
      callback(data);
    });
  }
}
```

앞의 함수는 cache 변수를 사용하여 서로 다른 파일을 읽어 작업의 결과를 저장한다. ~~위는 예제일 뿐, 오류 관리도 없고, 캐싱 로직 자체가 꼭 이렇게 되는 것은 아니다.~~ 
이 함수는 fs.readFile() 함수가 결과를 반환할 때까지 캐시가 설정되지 않은 경우 비동기식으로 동작하고, 캐시에 이미 있는 파일에 대한 모든 후속 요청에 대해 동기식으로 변해 즉각적으로 콜백을 호출하므로 위험하다.

### **Zalgo를 풀어놓다**

앞에서 정의한 것과 같이 예측할 수 없는 함수를 사용하면 어플리케이션이 쉽게 손상될 수 있다는 것을 살펴보자.

```javascript
function createFileReader(filename) {
  const listeners = [];
  inconsistentRead(filename, value => {
    listeners.forEach(listener => listener(value));
  });

  return {
    onDataReady: listener => listeners.push(listener)
  };
}
```

앞의 함수가 실행되면 파일 읽기에 대한 여러 가지 리스너(listener)를 설정할 수 있는, 알림을 발생시키는(notifier) 역할의 새로운 객체를 생성한다. 읽기가 완료되어 데이터 준비가 끝나면 모든 리스터들이 한번에 호출된다. 위 함수는 앞서 만든 inconsistentRead() 함수를 사용하여 이 기능을 구현한다. 이제 createFileReader() 함수를 사용해보자.

```javascript
// 테스트를 위해 data.txt를 만들고 파일 내에 some data 라고 작성해둠
const reader1 = createFileReader('data.txt'); 
reader1.onDataReady(data => {
  console.log('First call data: ' + data);

  //... 잠시 후 동일한 파일에 대해 다시 읽기를 수행한다.
  const reader2 = createFileReader('data.txt');
  reader2.onDataReady( data => {
    console.log('Second call data : ', data);
  })
})
```

이 코드는 다음과 같이 출력한다.

```javascript
First call data: some data
```

출력에서 볼 수 있듯이 두번째 콜백이호출되지 않는다. 그 이유는 아래와 같다.

- reader1이 생성되는 동안 inconsistentRead() 함수는 사용 가능한 캐시된 결과가 없으므로 비동기적으로 동작한다. 따라서 우리는 리스너를 등록하는데 충분한 시간을 가질 수 있다. 읽기 작업이 완료된 후, 나중에 이벤트 루프의 다른 사이클에서 리스너가 호출되기 때문이다.
- 그런 다음, reader2는 요청된 파일에 대한 캐시가 이미 존재하는 이벤트 루프의 사이클에서 생성된다. 이 경우는 inconsistentRead()에 대한 내부 호출은 동기 방식이 된다. 따라서 콜백은 즉시 호출된다. 즉, reader2의 모든 리스너들이 동기적으로 호출된다. 하지만, 리스너를 reader2의 생성 후에 등록하기 때문에 이들이 호출되는 일은 발생하지 않는다.

inconsistentRead() 함수의 콜백 동작은 실제로 호출 빈도, 인자로 전달되는 파일명 및 파일을 읽어들이는데 걸리는 시간과 같은 여러 요인에 의해 달라지므로 실제로 예측할 수 없다.

npm을 만들었고 Node.js 프로젝트의 리더인 Isaac Z. Schlueter는 이렇나 유형의 예측할 수 없는 함수들을 Zalgo를 풀어놓는(unleashing Zalgo)다고 비유했다. 

Zalgo는 세계의 광기, 죽음, 파괴를일으키는 것으로 믿어지는 불길한 존재에 대한 인터넷 상의 전설이다.

### 동기 API의 사용

Zalgo의 사례에서 알 수 있듯, API의 동기 또는 비동기 특성을 명확하게 정의하는 것이 필수적이다.

inconsistentRead()함수를 적절하게 수정할 수 있는 방법 중 한 가지는 완전히 동기화시키는 것이다. 이것은 Node.js가 대부분의 기본I/O 작업에 대한 동기식 직접 스타일 API 세트를 제공하기 때문에 가능하다. 예를 들어 비동기 형식 대신 fs.readFileSync() 함수를 사용할 수 있는데, 코드는 아래와 같다.

```javascript
const fs = require('fs');
const cache = {};
function consistentReadSync(filename){
  if(cache[filename]){
    callback(cache[filename]);
  }else {
    cache[filename] = fs.readFileSync(filename, 'utf8');
    return cache[filename];
  }
}
```

전체 기능이 직접 스타일로 변환되었다. 함수가 동기식이면 연속 전달 방식을 가질 이유가 없다. 실제로 직접 스타일을 사용하여 동기식 API를 구현하는 것이 항상 최선의 방법이라고 말할 수 있다. 이는 어플리케이션을 둘러싼 환경의 혼란을 제거하고 성능 측면에서 보다 효율적일 것이다. 

``순수한 동기식 함수에 대해서는 직접 스타일을 사용해라``

CPS에서 직접 스타일로 혹은 비동기에서 동기로 또는 그 반대로 API를 변경하면 API를 사용하는 모든 코드의 스타일을 변경해야 할 수도 있다. 일례로, 위의 경우 createFileReader() API의 인터페이스를 완전히 변경하고 항상 동기적으로 동작하도록 수정해야 한다.

또한, 비동기 API 대신 동기 API를 사용하면 몇 가지 주의해야 할 사항이 있다.

- 특정 기능에 대한 동기식 API를 항상 사용할 수 있는 것은 아니다.
- 동기 API는 이벤트 루프를 블록하고 동시 요청을 보류한다. JavaScript 동시성 모델을 깨뜨려서 전체 어플리케이션 속도를 떨어뜨린다.

앞서 consistentReadSync() 함수에서 동기식 I/O API는 하나의 파일당 한번 호출되고 이후의 호출에는 캐시에 저장된 값을 사용하기 때문에, 이벤트 루프를 블로킹하는 위험은 부분적으로 완화된다. 제한된 수의 정적 파일로 작업할 경우에는 consistentReadSync()를 사용하는 것은 이벤트 루프에 큰 영향을 미치지 않는다. 그러나 한 번만이라도 큰 파일을 읽는 경우라면 이야기는 완전히 달라진다. Node.js에서 동기 I/O를 사용하는 것은 많은 경우에 권장되지 않는다. 그렇다 하더라도, 어떤 경우에는 그것이 가장 쉽고 효율적인 해결안일 수 있다 적절한 대안을 선택하기 위해서는 항상 당면한 유스케이스를 먼저 살펴봐야 한다. 어플리케이션이 부팅되는(bootstrapping) 동안 동기 차단 API를 사용하여 환경 파일들을 로드(load)하는 것이 최적이다.

어플리케이션이 동시 요청을 처리하는데 영향을 주지 않는 경우에만 블로킹 API를 사용하라.

###지연 실행(Deferred execution)

inconsistentRead() 함수를 수정하는 또 다른 방법은 완전한 비동기로 만드는 것이다. 여기서 트릭은 동기 콜백 호출이 동일한 이벤트 루프 사잉클에서 즉시 실행되는 대신 "가까운 미래에" 실행되도록 예약하는 것이다. Node.js는 process.nextTick()을 사용하여 이 작업을 수행할 수 있다. process.nextTick()은 이벤트 루프의 다음 사이클까지 함수의 실행을 지연시킨다. 그 기능은 매우 간단하다. 콜백을 인수로 취하여 대기 중인 I/O 이벤트 대기열의 앞으로 밀어 넣고 즉시 반환한다. 그러면 콜백은 이벤트 루프가 다시 실행되는 즉시 호출한다.

이 기술을 적용하여 inconsistentRead() 함수를 다음과 같이 수정한다.

```javascript
const fs = require('fs');
const cache = {};
function consistentReadSync(filename){
  if(cache[filename]){
    process.nextTick(() => callback(cache[filename]));
  }else {
    //비동기함수
    fs.readFile(filename, 'utf8', (err, data) => {
      cache[filename] = data;
      callback(data);
    });  
  }
}
```

이 함수는 이제 어떤 상황에서도 콜백을 비동기적으로 호출할 수 있게 되었다. 

코드의 실행을 지연시키는 또 다른 API는 setImmediate()이다. 이 함수의 목적은 매우 유사하지만 그 의미가 크게 다르다. process.nextTick()으로 지연된 콜백은 다른 I/O 이벤트가 발생하기 전에 실행되지만, setImmediate()은 이미 큐에 있는 I/O 이벤트들의 뒤에 대기하게 된다. process.nextTick()은 이미 예정된 I/O보다 먼저 실행되기 때문에 재귀 호출과 같은 특정 상황에서 I/O 기아(starvation)를 발생시킬 수 있다.setImmediate()에서는 이런 일이 일어나지 않는다. 

``process.nextTick()을 사용하여 실행을 연기함으로써 콜백이 비동기적 호출을 보장할 수 있다.``

###2.1.3 Node.js 콜백 규칙

Node.js에서 연속 전달 스타일(CPS)의 API 및 콜백은 일련의 특정한 규칙을 따른다. 이 규칙은 Node.js 코어 API에 적용되지만 대다수의 사용자 영역 모듈과 어플리켕션에도 적용된다. 따라서 비동기 API를 설계할 때마다 이를 이해하고 반드시 준수해야 한다.

**콜백은 맨 마지막에**

모든 코어 Node.js 함수에서 표준 규칙은 함수가 입력에서 콜백을 허용한다면 맨 마지막 인자로 전달되어야 한다는 것이다. 다음 Node.js 코어 API를 예로 들어보자.

```javascript
fs.readFile(filename, [options], callback)
```

이 함수의 특성에서 볼 수 있듯이 여러 인자가 있는 경우에도 콜백은 항상 마지막 위치에 놓인다. 이 규칙의 이유는 콜백이 적절한 위치에 정의되어 있는 경우, 함수 호출의 가독성이 더 좋기 때문이다.

**오류는 맨 앞에**

CPS에서는 오류가 다른 유형의 결과처럼 전달되므로 콜백 사용이 필요하다. Node.js에서 CPS 함수에 의해 생성된 오류는 항상 콜백의 첫 번째 인수로 전달되며, 실제 결과는 두 번째 인수에서부터 전달된다. 동작이 에러 없이 성공하면, 최초의 인수는 null 혹은 undefine이 된다.  다음 코드는 이 규칙을 준수하는 콜백을 정의하는 방법을 보여준다.

```javascript
fs.readFile('foo.txt', 'utf8', (err, data) => {
  if(err)
    handleError(err);
  else
    processData(data);
});
```

에러가 있는지 항상 체크하는 것이 좋다. 그렇지 않으면 코드를 디버깅하고 에러 지점을 찾는 것이 어려울 수 있다. 고려해야할 또 다른 중요한 규칙은 오류는 항상 Error 유형(type)이어야 한다는 것이다. 즉, 간단한 문자열이나 숫자를 오류 객체로 전달해서는 안된다.

**오류 전파**

동기식 직접 스타일 함수의 오류 전파는 잘 알려진 throw 문을 사용하여 수행되므로 오류가 catch 될 때까지 호출 스택에서 실행된다.

그러나 비동기 CPS에서 적절한 오류 전달은 오류를 호출 체인의 다음에서 콜백으로 전달하여 수행된다. 일반적인 패턴은 다음과 같다.

```javascript
const fs = require('fs');
function readJSON(filename, callback) {
  fs.readFile(filename, 'utf8', (err, data) => {
    let parsed;
    if(err) // 오류를 전달하고 현재 함수를 종료
      return callback(err);

    try {
      // 파일의 내용을 해석
      parsed = JSON.parse(data);
    }catch(err) {
      // 에러를 catch
      return callback(err);
    }
    // 에러가 없으면 데이터를 전달
    callback(null, parse);
  })
}
```

이 코드에서 알아야 할 세부적인 사항은 유효한 결과를 전달할 때와 오류를 전달할 때의 콜백이 호출되는 방식이다. 또한 에러를 전파할 때 return 문을 사용한다는 것을 알아둬야 한다. 콜백 함수가 호출되는 즉시 함수에서 빠져 나와 readJSON 내의 그 다음 줄을 실행하지 않도록 한다.

**캐치되지 않는 예외**

readJSON 함수에서 보았듯이, fs.readFile() 함수 내 콜백에서 던져지는 예외를 피하기 위해 JSON.parse()를 try … catch 블록으로 둘러쌌다. 비동기 콜백 내부에서 예외를 발생시키면 예외가 이벤트 루프로 이동하여 다음 콜백으로 전파되지 않는다. Node.js에서 이것은 회복 불능의 상태이며, 어플리케이션은 그냥 종료되고 stderr 인터페이스를 통해 오류를 출력한다. 이를 재현해보기 위해 앞서 readJSON() 함수에서 정의했던 try … catch 구문을 제거해보면 다음과 같다.

```javascript
const fs = require('fs');
function readJSONThrows(filename, callback) {
  fs.readFile(filename, 'utf8', (err, data) => {
    if(err) {
     return callback(err);
    }
    // 에러가 없이 데이터만 전달
    callback(null, JSON.parse(data));
  });
};
```

이제 방금 정의한 함수에서는 JSON.parse()에서 발생하는 예외를 잡을 방법이 없다. 다음 코드로 잘못된 JSON 파일을 구문 분석해보자.

```javascript
readJSONThrows('nonJSON.txt', err => console.log(err));
```

어플리케이션이 갑작스럽게 종료되고 콘솔에 다음과 같은 예외 메시지가 출력된다.

```
SyntaxError: Unexpected token s
    at Object.parse (native)
    at [...]
    at tryToString (fs.js:457:3)
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:444:12)
```

앞의 스택 트레이스(stack trace)를 살펴보면 fs.js 모듈에서 시작하여, 정확히는 네이티브 API가 읽기를 완료한 후 이벤트 루프를 통해 fs.readFile() 함수로 그 결과를 반환한 지점으로부터 시작된다. 이것은 분명히 예외가 콜백에서 스택으로 이동한 다음, 즉시 이벤트 루프로 이동하여 마지막으로 콘솔에서 캐치(catch)되어 throw된다는 것을 보여준다.

이것은 readJSONThrows()를 try … catch 블록으로 둘러싸서 호출한다고 하더라도 블록이 동작하는 스택이 콜백이 호출된 스택과 다르기 때문에 동작하지 않는다는 것을 의미한다. 다음 코드는 방금 설명한 것처럼 이를 회피하기 위한 안티 패턴의 호출을 보여준다.

```javascript
try {
  readJSONThrows('nonJSON.txt', function(err, result){
    // ...
  });
} catch(err) {
  console.log('This will not catch the JSON parsing exception');
}
```

위 catch 문은 JSON 구문 분석의 예외를 받지 못한다. 예외가 발생한 스택과 실행 스택이 다르기 때문이다. 비동기 함수의 실행은 이벤트 루프에 의해 각기 다른 스택에서 실행되기 때문에 트리거 함수가 아닌 이벤트 루프에서 끝난다.

앞서 설명한 것처럼, 어플리케이션은 예외가 이벤트 루프에 도착하는 순간 중단된다. 그렇다 하더라도, 어플리케이션이 중단되기 전에 자원을 정리하거나 로그를 남길 수는 있다. 실제 이런 경우가 발생하면 Node.js는 프로세스를 종료하기 직전에 uncaughtException이라는 특수 이벤트를 내보낸다. 다음 코드는 이러한 경우 사용하는 코드의 예시이다.

```javascript
process.on('uncaughtException', (err) => {
  console.error('This will catch at last the' +
    'JSON parsing exception: ' + err.message);
  //종료 코드
});
```

캐치되지 않는 예외가 어플리케이션의 일관성을 보장할 수 없는 상태로 만든다. 이로 인해 예기치 않은 문제가 발생할 수 있음을 이해하는 것이 중요하다. 예를 들어, 여전히 불완전한 I/O 요청이 실행 중이거나 클로저가 일치하지 않을 수 있다. 따라서, 어쨋든 잡히지 않은 예외가 수신된 후, 특히 실제 운영 환경에서는 항상 어플리케이션을 종료하는 것이 좋다.